Skip reporting fault when the inner-most exception is OperationCanceledException

maryamariyan opened the pull request:
Skips fault reporting when the inner-most exception is `OperationCanceledException`.

Resolves: [WorkItem#1774918](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1774918/)

maryamariyan commented on src/Razor/test/Microsoft.VisualStudio.Editor.Razor.Test/TelemetryReporterTests.cs:
```diff
@@ -322,4 +322,40 @@ public void TrackLspRequest()
+            return new OperationCanceledException(message, innerException);
+        }
+
+        throw new NotImplementedException();
```
I am completing the implementation of this now based on what will be needed for https://github.com/dotnet/razor/pull/9919

davidwengier commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Telemetry/TelemetryReporter.cs:
```diff
@@ -109,6 +109,10 @@ public void ReportFault(Exception exception, string? message, params object?[] @
                 ReportFault(oceInnerException, message, @params);
                 return;
             }
+            else if (exception is OperationCanceledException)
```
I would combine this with the check above it. ie

```
if (exception is OperationCanceledException oce)
{
	// We don't want to report operation canceled, but don't want to miss out if there is something useful inside it
	if (oce.InnerException is not null)
	{
		ReportFault(oce, message, @params);
	}
	return;
}

maryamariyan commented on src/Razor/test/Microsoft.VisualStudio.Editor.Razor.Test/TelemetryReporterTests.cs:
```diff
@@ -322,4 +322,108 @@ public void TrackLspRequest()
+    }
+
+    [Theory]
+    [InlineData(3)]
```
FYI, would throw stackoverflow exception when depth is larger

maryamariyan commented on src/Razor/test/Microsoft.VisualStudio.Editor.Razor.Test/TelemetryReporterTests.cs:
```diff
@@ -322,4 +322,108 @@ public void TrackLspRequest()
+    [InlineData(typeof(RemoteInvocationException), 3)]
+    [InlineData(typeof(OperationCanceledException), 3)]
+    [InlineData(typeof(AggregateException), 15000)]
+    public void ReportFault_NestedException_NoStackOverflowException(Type exceptionType, int depth)
```
FYI, `ReportFault_NestedException_NoStackOverflowException` would throw stackoverflow exception for nested types `RemoteInvocationException` and `OperationCanceledException` when depth is larger

maryamariyan commented on src/Razor/test/Microsoft.VisualStudio.Editor.Razor.Test/TelemetryReporterTests.cs:
```diff
@@ -322,4 +322,108 @@ public void TrackLspRequest()
+    [InlineData(typeof(RemoteInvocationException), 3)]
+    [InlineData(typeof(OperationCanceledException), 3)]
+    [InlineData(typeof(AggregateException), 15000)]
+    public void ReportFault_MixedTypeNestedExceptions_NoStackOverflowException(Type exceptionType, int depth)
```
FYI, `ReportFault_MixedTypeNestedExceptions_NoStackOverflowException` would throw stackoverflow exception for nested types RemoteInvocationException and OperationCanceledException when depth is larger

maryamariyan commented:
All the tests I added here should pass as-is... but I added comments on PR explaining what input could cause them to crash with stackoverflow exception. (Related to https://github.com/dotnet/razor/pull/9919). 

I don't know if we want to fix them or if we think they may ever happen in the real world too. If not then I will just close PR https://github.com/dotnet/razor/pull/9919

davidwengier reviewed (commented):
Code fix itself is good, but I think the test fits better in the other PR than this one. Seems like there are three things to test with this change:
* That we don't report OperationCanceledException if there is no InnerException (the actual issue in the old code)
* That we don't report the InnerException of an OCE if it's another OCE (the old code, to ensure we don't regress)
* That we DO report the InnerException of an OCE if it's not an OCE (the desired behaviour, again to ensure we don't regress)

There is only test coverage for the middle one of these. Also it would be good to have a set of tests for TaskCanceledException too, to make sure inheritance is handled.

davidwengier commented on src/Razor/test/Microsoft.VisualStudio.Editor.Razor.Test/TelemetryReporterTests.cs:
```diff
@@ -322,4 +323,107 @@ public void TrackLspRequest()
+        // Arrange
+        var reporter = new TestTelemetryReporter(new RazorLoggerFactory([]));
+        var innerMostException = new OperationCanceledException();
+        var exception = CreateNestedException(typeof(OperationCanceledException), depth, innerMostException);
```
Nit: I would get rid of the helper methods and simplify this. This line could be replace with `new OperationCanceledException("", innerException: new OperationCanceledExcept())` and the test would still be valid.

Even having a nesting depth of 3 was particularly useful, having a theory with one data point, and two helper methods to create the structure, makes it harder to reason about what is going on and what the final shape of the exceptions looks like. Simply writing out the code directly using the constructors would make it much clearer IMO.

maryamariyan replied:
Good point! thanks for the feedback

davidwengier commented on src/Razor/test/Microsoft.VisualStudio.Editor.Razor.Test/TelemetryReporterTests.cs:
```diff
@@ -322,4 +323,107 @@ public void TrackLspRequest()
+
+    [Theory]
+    [InlineData(3)]
+    public void ReportFault_InnerMostExceptionIsOperationCanceledException_SkipsFaultReport(int depth)
```
This test looks to be identical to `ReportFault_InnerExceptionOfOCEIsAnotherOCE_SkipsFaultReport` above

davidwengier reviewed (approved):
Sorry, I'm bikeshedding about tests 😁

davidwengier reviewed (approved):
Sorry, I'm bikeshedding about tests 😁
